### PR TITLE
Remodel storage constraint ddl to reflect reality

### PR DESCRIPTION
**Pull request by wallyworld**
Here we rework the storage constraints DDL to reflect the implementation requirements.
The storage constraints are tied to an application or unit, not storage instance. The attributes are known, fixed, and fully populated for each constraint so we can model using a single table, one row per constraint.

Storage constraints are modelled similar to cloud or model config key values. The possible storage names come from the charm metadata; each application/unit has a tuple (pool, size, count) for each named storage. Thus the primary key of the app/unit storage constraint table is (parent uuid, storage name). The remaining columns are the constraint attributes (pool, size, count).

Charm storage tables are added to provide a parent record for the FK from the storage constraints table.

We store storage pool as a text value rather than a FK to the storage pool table because it can either be a pool OR a storage type (eg loop or ebs or tmps). The type can be a built in type or a provider specific type. Hence for the same reason we model storage pool type as a text value, we do so here too.

In line with the above, the storage_instance_pool table is removed and a text storage_pool attribute in the storage instance table is used instead.

## QA steps

Unit tests

**Jira card:** JUJU-5672

### CODE REVIEW

File: domain/schema/model.go [hunk 1]
```diff
@@ -558,11 +561,87 @@ INSERT INTO storage_kind VALUES
...
+-- Note that this is not unique; it speeds access by unit.
+CREATE INDEX idx_unit_storage_constraint
+ON unit_storage_constraint (unit_uuid);
+
 CREATE TABLE storage_instance (
     uuid            TEXT PRIMARY KEY,
     storage_kind_id INT NOT NULL,
     name            TEXT NOT NULL,
     life_id         INT NOT NULL,
+    -- Note: one might wonder why storage_pool below is not a
```
Comment (manadart) [inline domain/schema/model.go:638]: Then we should have nullable columns for type ID and pool ID, with the appropriate FKs.

File: domain/schema/model.go [hunk 2]
```diff
@@ -558,11 +561,87 @@ INSERT INTO storage_kind VALUES
...
+-- This table stores storage constraint values for each named storage item
+-- defined by the unit's current charm. If the charm is updated, then
+-- so too will be the rows in this table to reflect the current charm's
+-- storage definitions.
+-- Note: usually we just get the storage constraints off the application
+-- but need to allow for a unit's charm to temporarily diverge from that
+-- of its application.
+CREATE TABLE unit_storage_constraint (
+    unit_uuid               TEXT NOT NULL,
+    storage_name            TEXT NOT NULL,
```
Comment (manadart) [inline domain/schema/model.go:606]: This looks like there should be a table for `charm_storage` with the names listed there and maintained on install/refresh.

File: domain/schema/model.go (hunk 1 above)
Comment (wallyworld) [reply domain/schema/model.go:681]: As discussed live, this follows the same paradigm as used for storage pool type, reasoning is included in inline comments.

File: domain/schema/model.go [hunk 3]
```diff
@@ -558,11 +599,92 @@ INSERT INTO storage_kind VALUES
     (0, 'block', 'Allows for the creation of raw storage volumes'), 
     (1, 'filesystem', 'Provides a hierarchical file storage system');
 
+-- This table stores storage constraint values for each named storage item
+-- defined by the application's current charm. If the charm is updated, then
+-- so too will be the rows in this table to reflect the current charm's
```
Comment (manadart) [inline domain/schema/model.go:604]: This doesn't really make sense to me. An application has one charm.

A _unit_ might be on an old version of the charm compared to the application, but today that is indicated by charm URL. Given that a charm URL uniquely identifies a charm today, I suppose we will probably have a new charm record (new UUID) that we change the application to, then the units progressively.

To me it makes more sense to store a set of constraints declared by the charm against the charm itself, have this table keyed just on the application, and build the constraints hierarchically.

File: domain/schema/model.go [hunk 4]
```diff
@@ -558,11 +599,92 @@ INSERT INTO storage_kind VALUES
...
+-- This table stores storage constraint values for each named storage item
+-- defined by the unit's current charm. If the charm is updated, then
+-- so too will be the rows in this table to reflect the current charm's
+-- storage definitions.
+-- Note: usually we just get the storage constraints off the application
+-- but need to allow for a unit's charm to temporarily diverge from that
+-- of its application.
+CREATE TABLE unit_storage_constraint (
+    unit_uuid    TEXT NOT NULL,
+    charm_uuid   TEXT NOT NULL,
```
Comment (manadart) [inline domain/schema/model.go:648]: A unit will have some link to its current charm already, most likely in the actual unit table. There's no need to store it here; it makes for confusing uniqueness possibilities.

Comment (wallyworld) [reply domain/schema/model.go:648]: This was done because you requested a FK to the charm storage row to which this constraint pertains. The FK is the (charm_uuid, storage name) tuple. Charm storage rows do not have a uuid - they are a child of a parent charm and are identified by their parent charm uuid and name.
The PR originally just had storage name and omitted the charm uuid.

File: domain/schema/model.go (hunk 3 above)
Comment (wallyworld) [reply domain/schema/model.go:604]: Storage constraints are tied to an application not a charm. The charm does not itself declare storage constraints. The charm defines the allowed set of storage names it supports, in the charm_storage table. We create storage constraints for each application which uses that charm - each app storage constraint is for a charm storage name from the charm storage table. Each application will have a different set of constraints.

File: domain/schema/model.go [hunk 5]
```diff
@@ -369,6 +369,44 @@ CREATE TABLE charm (
...
+ON charm_storage (charm_uuid);
+
+CREATE TABLE charm_storage_property (
+    charm_uuid   TEXT NOT NULL,
+    storage_name TEXT NOT NULL,
+    property     TEXT NOT NULL,
+    CONSTRAINT       fk_charm_storage_property_charm_storage
+        FOREIGN KEY  (charm_uuid, storage_name)
+        REFERENCES   charm_storage(charm_uuid, name),
+    PRIMARY KEY (charm_uuid, storage_name, property)
```
Comment (manadart) [inline domain/schema/model.go:403]: What is an example of a property value?

File: domain/schema/model.go (hunk 3 above)
Comment (manadart) [reply domain/schema/model.go:604]: Isn't minimum size against the charm storage effectively a default constraint?

File: domain/schema/model.go (hunk 4 above)
Comment (manadart) [reply domain/schema/model.go:648]: I get this now. 

I would be inclined to have a single primary key on `charm_storage` (just an auto-incrementing ID if you don't want to model it as an outright entity), then the FK for `charm_storage_property`, `application_storage_constraint` and `unit_storage_constraint` is a single instead of composite key.

**Review by manadart — COMMENTED** [review]
Getting there.

File: domain/schema/model.go (hunk 4 above)
Comment (wallyworld) [reply domain/schema/model.go:648]: This would then model this 1..N differently than we do for other things where the N side has a composite primary key of the parent record and the unique name. If we are happy with that divergence in how we model these things, I can do that an auto inc ID but really prefer consistency. Yes, the approach here does in effect dup the charm_uuid, but only transitively. IMO it's an acceptable trade off for modelling consistency.
It's definitely not an outright entity; it's the same as for example a storage pool attribute where the composite PK is pool_uuid and attr name; storage pool attribute is not an entity in its own right, it doesn't get a uuid; it's part of a storage pool and doesn't exist on its own.

File: domain/schema/model.go (hunk 5 above)
Comment (wallyworld) [reply domain/schema/model.go:403]: The doc I found says
```
	// Properties allow the charm author to characterise the relative storage
	// performance requirements and sensitivities for each store.
	// eg “transient” is used to indicate that non persistent storage is acceptable,
	// such as tmpfs or ephemeral instance disks.
	//
	// Properties has no default, and is optional.
```
I didn't realise this existed

File: domain/schema/model.go (hunk 3 above)
Comment (wallyworld) [reply domain/schema/model.go:604]: Minimum size can be considered like that but it's not what "storage constraints" end up being. Actually, we really should call the juju model artefact "storage directive" and the table becomes "storage_directive". This is the correct terminology that we should be using to refer to this aspect. We talk about deployment with "--storage" directives. The allowed storage directive is informed / constrained by the rules in charm metadata (min/max count aka range, min size) but ultimately specified by the user with model / storage kind fallbacks. So the charm metadata constrains "pgdata" storage to be min 10GB and between 1 and 3 attached volumes. The user deploys postgresql with storage directive "--storage ebs,20GB,2". This falls within the bounds of the charm specified constraints so that means every unit of postgresql is realised with that exact storage requirement. Note that the charm does not specify the provider type (eg ebs), just the kind. The exact type is supplied as part of the storage directive by the user, so long as it matches the storage kind it's allowed.

So TL;DR; we should call this table "storage_directive". Then I hope the modelling of t makes more sense.

**Review by manadart — APPROVED** [review]
I can live with the tradeoffs here, but please let's do the storage _directive_ renaming. That makes more sense.

File: domain/schema/model.go (hunk 5 above)
Comment (manadart) [reply domain/schema/model.go:403]: Hmm. It's mentioned in https://juju.is/docs/sdk/storage. If it's not factored Juju-side, I bet no one is relying on it charm-side.

Comment (wallyworld) [reply domain/schema/model.go:403]: I say we leave it until we know it's needed.
